service/dap/types.go

```diff
@@ -0,0 +1,116 @@
+package dap
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+)
+
+// LaunchConfig is the collection of launch request attributes recognized by delve DAP implementation.
+type LaunchConfig struct {
+	// Command line arguments passed to the debugged program.
+	Args []string `json:"args,omitempty"`
+
+	// Backend used by delve. Maps to `dlv`'s `--backend` flag.
+	// Allowed Values: `"default"`, `"native"`, `"lldb"`
+	Backend string `json:"backend,omitempty"`
+
+	// Build flags, to be passed to the Go compiler. Maps to dlv's `--build-flags` flag.
+	BuildFlags string `json:"buildFlags,omitempty"`
+
+	// Absolute path to the working directory of the program being debugged
+	// if a non-empty value is specified. If not specified or empty,
+	// the working directory of the delve process will be used.
+	Cwd string `json:"cwd,omitempty"`
+
+	// Environment variables passed to the program.
+	Env map[string]string `json:"env,omitempty"`
+
+	// One of `debug`, `test`, `exec`.
+	Mode string `json:"mode,omitempty"`
+
+	// Output path for the binary of the debugee.
+	Output string `json:"output,omitempty"`
+
+	// Path to the program folder (or any go file within that folder)
+	// when in `debug` or `test` mode, and to the pre-built binary file
+	// to debug in `exec` mode.
+	// If it is not an absolute path, it will be interpreted as a path
+	// relative to the working directory of the delve process.
+	Program string `json:"program,omitempty"`
```

Comment:
maybe required fields shouldn't be omitempty?
OTOH, I think it's ok to rely on checking zero value in most cases.
(StackTraceDepth was a concern but the current code already ignores '0')

Reply:
I don't see the need of distinguishing "" and omitted value in case of replay mode either. 
So, I am voting for just treating zero value and unset value in the same way, and leaving this as omitempty except fields we are sure we know omitting and "" need to be handled differently.

Let me know if there are fields where omitted value and zero value need to be interpreted differently.
(Is it catastrophic to handle substitutePath's missing from/to fields as if they were set to ""?)

Reply:
I have no issues with having program as omitempty. Just wanted to highlight that it is not always a required field, so no special treatment is needed.

SubstitutePath of "" is not very useful. If omitted or empty, the mapping will be equally useless.

I am not sure about cwd. I guess we are free to define it whichever way we see fit. But as a point of reference, I think --wd of "" is a valid input too the backend with rpc setup.